SCClassLibrary/Common/Streams/EventStreamCleanup.sc

```diff
@@ -1,47 +1,86 @@
+// The "new" cleanup framework: built-in mechanism to avoid duplicate
+// execution of cleanups, difficult to avoid otherwise due to their
+// replicated nature in a stream graph. Instead of storing bare Functions,
+// they are now stored in CleanupThunks.
+
+// Unlike Thunks which take no arguments, CleanupThunks do allow arguments
+// to be passed to the function, but the CleanupThunk contract is that any
+// subsequent calls into the function would monotonically do the same or
+// less (cleanup) work, so that the first call does the maximal amount of
+// relevant work, making subsequent calls completely redundant.
+// Prototypical example: if a cleanup is called with a "flag: false",
+// it is assumed it will *not* be called later again with "flag: true",
+// if the latter means that the function would have to do "extra" work
+// that was skipped/excluded by "flag: false".
+// CleanupThunk is not a general-purpose function memoization class for
+// functions with arguments.
+CleanupThunk : Thunk {
```

Comment:
Given that this is not a general-purpose class, but also not a class that is limited to cleanup, it would be worth giving it a name thatdescribes its function outside the context of cleanup?

Reply:
Well, I'm open to naming suggestions... `CallOnce`, which you suggested in the bug discussion seemed a bit too contrived to indicate it differed from Thunk...

Reply:
If we really think of it, the purpose is quite different. 
- A Thunk is a way to abstract from the difference between a function and its result (this is why it has no arguments). 
- The object we need is something that limits the number of times a function can be called. 

Somehow this is more like a stream, like `Pstutter(inf, Pfunc(func)).asStream`, internally it looks more like a `FuncStream`.

I am tempted to suggest `ThunkStream`.

Reply:
This is actually an interesting perspective as cleanups are also produced and consumed. I'll have to ponder on it a bit more before I say more though.

Reply:
While finalizers in general can be considered [out-of-band data](https://en.wikipedia.org/wiki/Out-of-band_data), I think calling the CleanupThunk a stream conflict with the narrower notion of stream used in SC because Streams don't just have  a `next` function but also an `emmbedInStream` which yields items. And CleanupThunks don't really have that. Nor can they even be considered Routines for the same reason that yield doesn't work in them in any special way. So, I think  CleanupThunk is properly and AbstractFunction (which it actually is in the hierarchy since Thunk subclasses AbstractFunction too).

I'm not too hung-up on this, but I think renaming CleanupThunk to a Stream would be misleading in this sense.